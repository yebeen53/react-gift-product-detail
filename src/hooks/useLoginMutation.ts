import { useMutation } from '@tanstack/react-query';
import { apiClient } from '@/api/apiClient';

interface LoginPayload {
  email: string;
  password: string;
}

interface LoginResponse {
  authToken: string;
  id: number;
  name: string;
  email: string;
}

export const useLoginMutation = () => {
  return useMutation({
    mutationFn: async (payload: LoginPayload): Promise<LoginResponse> => {
      const res = await apiClient.post('/api/login', payload);
      return res.data.data;
    },
  });
};