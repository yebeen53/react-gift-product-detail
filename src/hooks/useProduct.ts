import {
  useSuspenseQuery,
  useMutation,
  useQueryClient,
} from '@tanstack/react-query';
import {
  fetchProduct,
  fetchProductDetail,
  fetchHighlightReview,
  fetchWishCount,
} from '@/api/products/product';

export const useProduct = (productId: string) =>
  useSuspenseQuery({
    queryKey: ['product', productId],
    queryFn: () => fetchProduct(productId),
  });

export const useProductDetail = (productId: string) =>
  useSuspenseQuery({
    queryKey: ['productDetail', productId],
    queryFn: () => fetchProductDetail(productId),
  });

export const useHighlightReview = (productId: string) =>
  useSuspenseQuery({
    queryKey: ['highlightReview', productId],
    queryFn: () => fetchHighlightReview(productId),
  });

export const useWishCount = (productId: string) =>
  useSuspenseQuery({
    queryKey: ['wishCount', productId],
    queryFn: () => fetchWishCount(productId),
  });

export const useWishMutation = (productId: string) => {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: async () => Promise.resolve(), // 실제 API 요청 함수 넣기

    onMutate: async () => {
      await queryClient.cancelQueries({ queryKey: ['wishCount', productId] });

      const prevWish = queryClient.getQueryData<{
        wishCount: number;
        isWished: boolean;
      }>(['wishCount', productId]);

      if (prevWish) {
        queryClient.setQueryData(['wishCount', productId], {
          ...prevWish,

          wishCount: prevWish.isWished
            ? prevWish.wishCount - 1
            : prevWish.wishCount + 1,
          isWished: !prevWish.isWished, // 토글
        });
      }

      return { prevWish };
    },

    onError: (_err, _vars, context) => {
      if (context?.prevWish) {
        queryClient.setQueryData(['wishCount', productId], context.prevWish);
      }
    },

    onSettled: () => {
      queryClient.invalidateQueries({ queryKey: ['wishCount', productId] });
    },
  });
};
