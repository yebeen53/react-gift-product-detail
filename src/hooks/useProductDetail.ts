import { useSuspenseQuery } from '@tanstack/react-query';
import axios from 'axios';

interface Announcement {
  name: string;
  value: string;
  displayOrder: number;
}

interface ProductDetail {
  description: string;
  announcements: Announcement[];
}

const fetchProductDetail = async (productId: string): Promise<ProductDetail> => {
  const response = await axios.get(`/api/products/${productId}/detail`);
  return response.data.data;
};

export const useProductDetail = (productId: string) =>
  useSuspenseQuery({
    queryKey: ['productDetail', productId],
    queryFn: () => fetchProductDetail(productId),
  });
