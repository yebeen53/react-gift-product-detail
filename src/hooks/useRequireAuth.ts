import useAuth from '@/hooks/useAuth';
import { useLocation, useNavigate } from 'react-router-dom';
import { useEffect } from 'react';
import { ROUTES } from '@/constants/routes';

const useRequireAuth = () => {
  const { user } = useAuth();
  const location = useLocation();
  const navigate = useNavigate();

  useEffect(() => {
    if (!user || !user.authToken) {
      navigate(ROUTES.HOME, {
        replace: true,
        state: { from: location.pathname },
      });
    }
  }, [user, navigate, location]);

  return user;
};

export default useRequireAuth;
