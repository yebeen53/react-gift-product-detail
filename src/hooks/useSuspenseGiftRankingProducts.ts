import { fetchRanking } from '@/api/products/ranking';
import { useSuspenseQuery } from '@tanstack/react-query';

export const useSuspenseGiftRankingProducts = (
  genderCode: string,
  categoryCode: string
) => {
  return useSuspenseQuery({
    queryKey: ['giftRankingProducts', genderCode, categoryCode],
    queryFn: () => fetchRanking(genderCode, categoryCode),
  });
};
