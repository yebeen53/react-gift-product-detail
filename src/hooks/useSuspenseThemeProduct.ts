import {
  useSuspenseInfiniteQuery,
  type QueryFunctionContext,
} from '@tanstack/react-query';
import axios from 'axios';
import type { ThemeProductsData } from '@/types/product';

const fetchThemeProducts = async (
  ctx: QueryFunctionContext<readonly [string, string], number>
): Promise<ThemeProductsData> => {
  const [, themeId] = ctx.queryKey;
  const pageParam = ctx.pageParam ?? 0;

  const res = await axios.get(`/api/themes/${themeId}/products`, {
    params: { cursor: pageParam, limit: 10 },
  });
  const data = res.data.data;

  return {
    list: data.list,
    cursor: data.cursor,
    hasMoreList: data.hasMoreList,
  };
};

export const useSuspenseThemeProducts = (themeId: string) => {
  return useSuspenseInfiniteQuery({
    queryKey: ['themeProducts', themeId] as const,
    queryFn: fetchThemeProducts,
    initialPageParam: 0,
    getNextPageParam: (lastPage) =>
      lastPage.hasMoreList ? lastPage.cursor : undefined,
  });
};
