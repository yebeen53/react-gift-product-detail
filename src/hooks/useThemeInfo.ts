import { apiClient } from '@/api/apiClient';
import { useSuspenseQuery } from '@tanstack/react-query';

export interface ThemeDetail {
  image: string | undefined;
  name: string;
  themeId: string;
  title: string;
  description: string | null;
  backgroundColor: string;
}

const fetchThemeDetail = async (themeId: string): Promise<ThemeDetail> => {
  const res = await apiClient.get(`/api/themes/${themeId}/info`);
  const data = res.data?.data;
  if (!data) {
    throw new Error('테마 정보가 없습니다.');
  }
  return data;
};

export const useThemeInfo = (themeId: string) => {
  return useSuspenseQuery({
    queryKey: ['themeDetail', themeId],
    queryFn: () => fetchThemeDetail(themeId),
  });
};
