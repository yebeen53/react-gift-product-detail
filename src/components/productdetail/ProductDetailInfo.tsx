import { Suspense } from 'react';
import { useProductDetail } from '@/hooks/useProduct';
import theme from '@/data/theme';

interface Props {
  productId: string;
}

const ProductDetailInfoInner = ({ productId }: Props) => {
  const { data: detail } = useProductDetail(productId);
  if (!detail) {
    return <div>로딩중...</div>; 
  }

  return (
    <section style={{ margin: theme.spacing.spacing4 }}>
      <h2>상품설명</h2>
      <div dangerouslySetInnerHTML={{ __html: detail.description }} />
      {detail.images?.map((img: string, idx: number) => (
        <img key={img} src={img} alt={`상품 상세 이미지${idx}`} />
      ))}
    </section>
  );
};

const ProductDetailInfo = (props: Props) => (
  <Suspense fallback={<div>Loading product detail...</div>}>
    <ProductDetailInfoInner {...props} />
  </Suspense>
);

export default ProductDetailInfo;
