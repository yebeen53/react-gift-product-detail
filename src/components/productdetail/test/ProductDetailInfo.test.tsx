import { render, screen, waitFor } from '@testing-library/react';
import ProductDetailInfo from '../ProductDetailInfo';
import { useProductDetail } from '@/hooks/useProduct';

jest.mock('@/hooks/useProduct', () => ({
  useProductDetail: jest.fn(),
}));

describe('ProductDetailInfo 컴포넌트', () => {
  const mockProductDetail = {
    description: '<p>이것은 <strong>상품 설명</strong>입니다.</p>',
    images: ['image1.jpg', 'image2.jpg'],
  };

  beforeEach(() => {
    (useProductDetail as jest.Mock).mockClear();
  });

  it('데이터 로딩 중에는 폴백 메시지가 렌더링되어야 한다', () => {
    (useProductDetail as jest.Mock).mockReturnValue({ data: undefined });
    const { container } = render(<ProductDetailInfo productId="p1" />);
    expect(container).toHaveTextContent(/로딩중/i);
  });

  it('상품 설명과 이미지가 올바르게 렌더링되어야 한다', async () => {
    (useProductDetail as jest.Mock).mockReturnValue({ data: mockProductDetail });
    render(<ProductDetailInfo productId="p1" />);

    await waitFor(() => {
      expect(screen.getByRole('heading', { name: /상품설명/i })).toBeInTheDocument();
      expect(screen.getByText(/이것은/)).toBeInTheDocument();
      expect(screen.getByText(/상품 설명/)).toBeInTheDocument();

      const images = screen.getAllByRole('img');
      expect(images).toHaveLength(2);
      expect(images[0]).toHaveAttribute('src', 'image1.jpg');
      expect(images[1]).toHaveAttribute('src', 'image2.jpg');
    });
  });

  it('컴포넌트의 스타일 및 구조가 스냅샷과 일치해야 한다', async () => {
    (useProductDetail as jest.Mock).mockReturnValue({ data: mockProductDetail });
    const { asFragment } = render(<ProductDetailInfo productId="p1" />);
    await waitFor(() => {
      expect(screen.getByText(/상품 설명/)).toBeInTheDocument();
    });
    expect(asFragment()).toMatchSnapshot();
  });
});
