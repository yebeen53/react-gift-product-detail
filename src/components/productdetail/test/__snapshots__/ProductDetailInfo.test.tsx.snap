// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ProductDetailInfo 컴포넌트 컴포넌트의 스타일 및 구조가 스냅샷과 일치해야 한다 1`] = `
<DocumentFragment>
  <section
    style="margin: 16px;"
  >
    <h2>
      상품설명
    </h2>
    <div>
      <p>
        이것은 
        <strong>
          상품 설명
        </strong>
        입니다.
      </p>
    </div>
    <img
      alt="상품 상세 이미지0"
      src="image1.jpg"
    />
    <img
      alt="상품 상세 이미지1"
      src="image2.jpg"
    />
  </section>
</DocumentFragment>
`;
